### PR TITLE
ONNX: modifications in onnx_importer for upsample node fused from unfused Resize subgraph

### LINKED ISSUE #17001: Cannot load the Midas model(model of pytorch) after convert to onnx
##### System information (version)

- OpenCV => 4.3.0(e6f32c6a69043456a806a4e802ee3ce7b7059c93)
- Operating System / Platform => Windows 64 Bit
- Compiler =>Visual Studio 2017 64 bits

##### Detailed description

The [model ](https://drive.google.com/file/d/1nqW_Hwj86kslfsXR7EnXpEWdO2csz1cC)convert from [this project](https://github.com/intel-isl/MiDaS) do not work.

##### Steps to reproduce

1. install pytorch 1.4.0(conda install pytorch torchvision opencv),  please use python=3.6 since onnx do not support python3.7
2. install onnx
3. clone the [project](https://github.com/intel-isl/MiDaS)
4. [Download the model](https://drive.google.com/file/d/1nqW_Hwj86kslfsXR7EnXpEWdO2csz1cC) and place it in the root folder of the project
5. Create to_onnx.py in the root folder

to_onnx.py

```
import os
import glob
import torch
import utils
import cv2
import numpy as np

from torchvision.transforms import Compose
from models.midas_net import MidasNet
from models.transforms import Resize, NormalizeImage, PrepareForNet

import onnx
import onnxruntime

def test_model_accuracy(export_model_name, raw_output, input):    
    ort_session = onnxruntime.InferenceSession(export_model_name)

    def to_numpy(tensor):
        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

    # compute ONNX Runtime output prediction
    ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(input)}
    ort_outs = ort_session.run(None, ort_inputs)	

    # compare ONNX Runtime and PyTorch results
    np.testing.assert_allclose(to_numpy(raw_output), ort_outs[0], rtol=1e-03, atol=1e-05)

    print("Exported model has been tested with ONNXRuntime, and the result looks good!")		

def export_model(model, input, export_model_name):
    torch.onnx.export(model, input, export_model_name, verbose=False, export_params=True, opset_version=11)	
    onnx_model = onnx.load(export_model_name)    
    onnx.checker.check_model(onnx_model)
    graph_output = onnx.helper.printable_graph(onnx_model.graph)
    with open("graph_output.txt", mode="w") as fout:
        fout.write(graph_output)
		
device = torch.device("cpu")

 # load network
model_path = "model.pt"
model = MidasNet(model_path, non_negative=True)

transform = Compose(
        [
            Resize(
                384,
                384,
                resize_target=None,
                keep_aspect_ratio=True,
                ensure_multiple_of=32,
                resize_method="lower_bound",
                image_interpolation_method=cv2.INTER_CUBIC,
            ),
            NormalizeImage(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
            PrepareForNet(),
        ]
)

model.to(device)
model.eval()

img = utils.read_image("input/line_up_00.jpg")
img_input = transform({"image": img})["image"]

# compute
#with torch.no_grad():
sample = torch.from_numpy(img_input).to(device).unsqueeze(0)
print("sample type = ", type(sample), ", shape of sample = ", sample.shape)
prediction = model.forward(sample.clone())

export_model_name = "midas.onnx"	

#encoder_script_module = torch.jit.trace(model, img_input)
#encoder_script_module.save(export_model_name)

export_model(model, sample, export_model_name)
test_model_accuracy(export_model_name, prediction, sample)
```

Load the model with following codes

```
std::string const model_folder("C:/Users/yyyy/programs/Qt/app_to_simplify_ml/"
                                   "cv_projects/clients/mobile_depth_estimation/MiDaS");
cv::dnn::Net net = cv::dnn::readNet(model_folder + "/midas.onnx");
```

Got error messages

**OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.3.0-pre) Error: Assertion failed (index < node_proto.input_size()) in cv::dnn::dnn4_v20200310::ONNXImporter::getBlob, file C:\Users\yyyy\programs\Qt\3rdLibs\opencv\dev\opencv\modules\dnn\src\onnx\onnx_importer.cpp, line 257**

For simplicity, you can have the converted model, graph and to_onnx.py from the [google drive](https://drive.google.com/open?id=1VNjrV2LJY8s_ieLSuBASetLVkelr_tTp).

### PR DESCRIPTION
**Pull request by ashishkrshrivastava**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/743

resolves #17001 

```
opencv_extra=onnximporter
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### This PR modifies
Parsing for `Upsample` node obtained from fusion of Unfused Resized Subgraph by adding setting of layer parameters.

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.1.0:16.04
build_image:Custom Win=openvino-2020.1.0
build_image:Custom Mac=openvino-2020.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -1355,8 +1355,7 @@ void ONNXImporter::populateNet(Net dstNet)
                 CV_Assert(layer_id.find(node_proto.input(i)) == layer_id.end());
 
             String interp_mode = layerParams.get<String>("coordinate_transformation_mode");
-            CV_Assert_N(interp_mode != "tf_crop_and_resize", interp_mode != "asymmetric",
-                        interp_mode != "tf_half_pixel_for_nn");
+            CV_Assert_N(interp_mode != "tf_crop_and_resize", interp_mode != "tf_half_pixel_for_nn");
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1358]: What is `asymmetric` mode and do we really support it?

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -1401,7 +1400,23 @@ void ONNXImporter::populateNet(Net dstNet)
...
+                    CV_Assert_N(interp_mode != "tf_crop_and_resize", interp_mode != "tf_half_pixel_for_nn");
+
+                    layerParams.set("align_corners", interp_mode == "align_corners");
+                    if (layerParams.get<String>("interpolation") == "linear") {
+                        layerParams.set("interpolation", interp_mode == "pytorch_half_pixel" ?
+                                                "opencv_linear" : "bilinear");
+                    }
+                }
+            }
+            else if (layerParams.get<String>("interpolation") == "linear" && framework_name == "pytorch") {
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1419]: Can you please refactor so we first extract interpolation (without checking number of inputs) but separately scale factors? So make it more similar to `else if (layer_type == "Resize") above.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 1 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_importer.cpp:1358]: https://github.com/onnx/onnx/blob/master/docs/Operators.md#resize

> if coordinate_transformation_mode is "asymmetric",
> x_original = x_resized / scale,

If I am not wrong we do support it. 

https://github.com/opencv/opencv/blob/5c7c80dd27abb24051097653823f7b8f22d8ac33/modules/dnn/src/layers/resize_layer.cpp#L78-L79

https://github.com/opencv/opencv/blob/5c7c80dd27abb24051097653823f7b8f22d8ac33/modules/dnn/src/layers/resize_layer.cpp#L83-L84

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_importer.cpp:1419]: okay.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 1 above)
Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:1358]: And we have test data for it in this PR, right?

Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_importer.cpp:1358]: https://github.com/opencv/opencv/blob/823fc5efb8b0347aff1e945276fecccf66c95557/modules/dnn/test/test_onnx_importer.cpp#L370-L371

These two test models has `coordinate_transformation_mode` parameter value `asymmetric`
So, yes, we have test data for `asymmetric`. 

> in this PR

This test is not added in this PR but was already there. Those tests were working fine before because those subgraphs were fused to Upsample so `coordinate_transformation_mode` was not getting checked and `asymmetric` is default for Resize layer in DNN module.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Please have a look if changes made are same as expected or you asked for something else.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 3]
```diff
@@ -1384,6 +1383,19 @@ void ONNXImporter::populateNet(Net dstNet)
         }
         else if (layer_type == "Upsample")
         {
+            //fused from Resize Subgraph
+            if (layerParams.has("coordinate_transformation_mode"))
```
Comment (ashishkrshrivastava) [inline modules/dnn/src/onnx/onnx_importer.cpp:1387]: I have to add this check because `coordinate_transormation_mode` parameter is not there in `Upsample` Node. So this `if` block will run only for `Resize` subgraph that is fused to `upsample`

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 4]
```diff
@@ -1384,31 +1383,46 @@ void ONNXImporter::populateNet(Net dstNet)
...
+
+                layerParams.set("align_corners", interp_mode == "align_corners");
+                if (layerParams.get<String>("mode") == "linear")
+                {
+                    layerParams.set("mode", interp_mode == "pytorch_half_pixel" ?
+                                            "opencv_linear" : "bilinear");
+                }
+            }
             layerParams.type = "Resize";
+            //scale as attribute
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1400]: I think we can emit this comment

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 5]
```diff
@@ -1384,31 +1383,46 @@ void ONNXImporter::populateNet(Net dstNet)
...
             if (layerParams.has("scales"))
             {
-                // Pytorch layer
+                // Pytorch
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1403]: Please keep unchanged

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 6]
```diff
@@ -1384,31 +1383,46 @@ void ONNXImporter::populateNet(Net dstNet)
...
             {
                 // Caffe2 layer
                 replaceLayerParam(layerParams, "height_scale", "zoom_factor_y");
                 replaceLayerParam(layerParams, "width_scale", "zoom_factor_x");
             }
-            replaceLayerParam(layerParams, "mode", "interpolation");
-
-            if (layerParams.get<String>("interpolation") == "linear" && framework_name == "pytorch") {
-                layerParams.type = "Resize";
+            //scale as input
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1415]: Please remove or move under `else`

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 4 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Done!

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 5 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Done!

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 6 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Done!
